src/promptflow/promptflow/_sdk/_service/utils/utils.py

```diff
@@ -37,44 +44,59 @@ def wrapper(*args, **kwargs):
+            service_config = load_yaml(f) or {}
+            port = service_config.get(sys.executable, {}).get("port", None)
+        if not port and create_if_not_exists:
             with open(HOME_PROMPT_FLOW_DIR / PF_SERVICE_PORT_FILE, "w", encoding=DEFAULT_ENCODING) as f:
-                service_config.pop("service")
+                # Set random port to ~/.promptflow/pf.yaml
+                port = get_random_port()
+                service_config[sys.executable] = service_config.get(sys.executable, {})
+                service_config[sys.executable]["port"] = port
                 dump_yaml(service_config, f)
+        return port
+    finally:
+        pfs_port_lock.release()
+
+
+def kill_service_get_from_service_field():
+    global pfs_port_lock
+    pfs_port_lock.acquire()
+    try:
+        (HOME_PROMPT_FLOW_DIR / PF_SERVICE_PORT_FILE).touch(mode=read_write_by_user(), exist_ok=True)
+        with open(HOME_PROMPT_FLOW_DIR / PF_SERVICE_PORT_FILE, "r", encoding=DEFAULT_ENCODING) as f:
+            service_config = load_yaml(f) or {}
+            port = service_config.get("service", {}).get("port", None)
+            if port:
+                if is_port_in_use(port):
+                    logger.debug(f"Kill the deprecated port {port} got from service key in thr pfs.port file.")
+                    kill_exist_service(port)
+                # remove service key field
+                with open(HOME_PROMPT_FLOW_DIR / PF_SERVICE_PORT_FILE, "w", encoding=DEFAULT_ENCODING) as f:
+                    service_config.pop("service")
+                    dump_yaml(service_config, f)
+    finally:
+        pfs_port_lock.release()
 
 
 def dump_port_to_config(port):
     # Set port to ~/.promptflow/pf.port, if already have a port in file , will overwrite it.
-    (HOME_PROMPT_FLOW_DIR / PF_SERVICE_PORT_FILE).touch(mode=read_write_by_user(), exist_ok=True)
-    with open(HOME_PROMPT_FLOW_DIR / PF_SERVICE_PORT_FILE, "r", encoding=DEFAULT_ENCODING) as f:
-        service_config = load_yaml(f) or {}
-    with open(HOME_PROMPT_FLOW_DIR / PF_SERVICE_PORT_FILE, "w", encoding=DEFAULT_ENCODING) as f:
-        service_config[sys.executable] = service_config.get(sys.executable, {})
-        service_config[sys.executable]["port"] = port
-        dump_yaml(service_config, f)
+    global pfs_port_lock
+    pfs_port_lock.acquire()
+    try:
+        (HOME_PROMPT_FLOW_DIR / PF_SERVICE_PORT_FILE).touch(mode=read_write_by_user(), exist_ok=True)
+        with open(HOME_PROMPT_FLOW_DIR / PF_SERVICE_PORT_FILE, "r", encoding=DEFAULT_ENCODING) as f:
+            service_config = load_yaml(f) or {}
+        with open(HOME_PROMPT_FLOW_DIR / PF_SERVICE_PORT_FILE, "w", encoding=DEFAULT_ENCODING) as f:
+            service_config[sys.executable] = service_config.get(sys.executable, {})
+            service_config[sys.executable]["port"] = port
+            dump_yaml(service_config, f)
+    finally:
+        pfs_port_lock.release()
 
 
 def is_port_in_use(port: int):
```

Comment:
what's the timeout for inner connect_ex() call?

Reply:
will fix in the next pr